src/PolygonClipper/StablePriorityQueue{T,TComparer}.cs

```diff
@@ -39,108 +47,124 @@ public StablePriorityQueue(TComparer comparer)
+            heapRef = default(T)!;
         }
 
-        return this.heap[0];
+        return top;
     }
 
     /// <summary>
     /// Restores the heap property by moving the item at the specified index upward.
     /// </summary>
     /// <param name="index">The index of the item to move upward.</param>
+    [MethodImpl(MethodImplOptions.AggressiveInlining)]
     private void Up(int index)
     {
-        List<T> data = this.heap;
-        T item = data[index];
+        ref T heapRef = ref MemoryMarshal.GetArrayDataReference(this.heap);
+        T item = Unsafe.Add(ref heapRef, index);
         TComparer comparer = this.Comparer;
 
         while (index > 0)
         {
             int parent = (index - 1) >> 1;
-            T current = data[parent];
-            if (comparer.Compare(item, current) >= 0)
+            ref T currentRef = ref Unsafe.Add(ref heapRef, parent);
+            if (comparer.Compare(item, currentRef) >= 0)
             {
                 break;
             }
 
-            data[index] = current;
+            Unsafe.Add(ref heapRef, index) = currentRef;
             index = parent;
         }
 
-        data[index] = item;
+        Unsafe.Add(ref heapRef, index) = item;
     }
 
     /// <summary>
     /// Restores the heap property by moving the item at the specified index downward.
     /// </summary>
     /// <param name="index">The index of the item to move downward.</param>
+    [MethodImpl(MethodImplOptions.AggressiveInlining)]
     private void Down(int index)
     {
-        List<T> data = this.heap;
-        int halfLength = data.Count >> 1;
-        T item = data[index];
+        ref T heapRef = ref MemoryMarshal.GetArrayDataReference(this.heap);
+        int halfLength = this.count >> 1;
+        T item = Unsafe.Add(ref heapRef, index);
         TComparer comparer = this.Comparer;
 
         while (index < halfLength)
         {
             int bestChild = (index << 1) + 1; // Initially left child
             int right = bestChild + 1;
 
-            if (right < data.Count && comparer.Compare(data[right], data[bestChild]) < 0)
+            if (right < this.count && comparer.Compare(Unsafe.Add(ref heapRef, right), Unsafe.Add(ref heapRef, bestChild)) < 0)
             {
                 bestChild = right;
             }
 
-            if (comparer.Compare(data[bestChild], item) >= 0)
+            ref T bestChildRef = ref Unsafe.Add(ref heapRef, bestChild);
+            if (comparer.Compare(bestChildRef, item) >= 0)
             {
                 break;
             }
 
-            data[index] = data[bestChild];
+            Unsafe.Add(ref heapRef, index) = bestChildRef;
             index = bestChild;
         }
 
-        data[index] = item;
+        Unsafe.Add(ref heapRef, index) = item;
+    }
+
+    /// <summary>
+    /// Resizes the internal array when capacity is exceeded.
+    /// </summary>
+    [MethodImpl(MethodImplOptions.NoInlining)]
+    private void Resize()
```

Comment:
We should be resizing using the same rules as `List<T>`
https://source.dot.net/#System.Private.CoreLib/src/libraries/System.Private.CoreLib/src/System/Collections/Generic/List.cs,488